src/chk/chk_leader.c

```diff
@@ -333,20 +335,1036 @@ chk_leader_find_slowest(struct chk_instance *ins)
+chk_leader_reset_pool_svc(struct chk_pool_rec *cpr)
+{
+	struct ds_pool_clues	*clues = &cpr->cpr_clues;
+	struct chk_instance	*ins = cpr->cpr_ins;
+	struct chk_bookmark	*cbk = &ins->ci_bk;
+	d_rank_t		*ranks;
+	d_rank_list_t		 rank_list;
+	d_iov_t			 psid;
+	int			 chosen = cpr->cpr_advice;
+	int			 i;
+	int			 j;
+	int			 rc;
+
+	D_ASSERT(chosen >= 0 && clues->pcs_len > chosen);
+
+	D_ALLOC_ARRAY(ranks, clues->pcs_len);
+	if (ranks == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	/* Build a list of all ranks except for the chosen one. */
+	for (i = 0, j = 0; i < clues->pcs_len; i++) {
+		if (i != chosen)
+			ranks[j++] = clues->pcs_array[i].pc_rank;
+	}
+
+	rank_list.rl_ranks = ranks;
+	rank_list.rl_nr = j;
+
+	d_iov_set(&psid, cpr->cpr_uuid, sizeof(uuid_t));
+	rc = ds_rsvc_dist_stop(DS_RSVC_CLASS_POOL, &psid, &rank_list, NULL /* excluded */,
+			       true /* destroy */);
+	D_FREE(ranks);
+
+out:
+	if (rc != 0) {
+		D_ERROR(DF_LEADER" failed to destroy other pool service replicas for "DF_UUIDF
+			": "DF_RC"\n", DP_LEADER(ins), DP_UUID(cpr->cpr_uuid), DP_RC(rc));
+
+		cbk->cb_statistics.cs_failed++;
+		cpr->cpr_skip = 1;
+	}
+
+	return rc;
+}
+
+static int
+chk_leader_dangling_pool(struct chk_instance *ins, uuid_t uuid)
+{
+	struct chk_property		*prop = &ins->ci_prop;
+	struct chk_bookmark		*cbk = &ins->ci_bk;
+	struct chk_report_unit		 cru;
+	Chk__CheckInconsistClass	 cla;
+	Chk__CheckInconsistAction	 act;
+	uint64_t			 seq = 0;
+	uint32_t			 options[2];
+	uint32_t			 option_nr = 0;
+	int				 decision = -1;
+	int				 result = 0;
+	int				 rc = 0;
+
+	cla = CHK__CHECK_INCONSIST_CLASS__CIC_POOL_NONEXIST_ON_ENGINE;
+	act = prop->cp_policies[cla];
+	cbk->cb_statistics.cs_total++;
+
+	switch (act) {
+	case CHK__CHECK_INCONSIST_ACTION__CIA_DEFAULT:
+		/*
+		 * Default action is to de-register the dangling pool from MS.
+		 *
+		 * Fall through.
+		 */
+	case CHK__CHECK_INCONSIST_ACTION__CIA_DISCARD:
+		/* Fall through. */
+	case CHK__CHECK_INCONSIST_ACTION__CIA_TRUST_PS:
+		seq = ++(ins->ci_seq);
+		if (prop->cp_flags & CHK__CHECK_FLAG__CF_DRYRUN) {
+			cbk->cb_statistics.cs_repaired++;
+		} else {
+			result = ds_chk_deregpool_upcall(seq, uuid);
+			if (result != 0)
+				cbk->cb_statistics.cs_failed++;
+			else
+				cbk->cb_statistics.cs_repaired++;
+		}
+		break;
+	case CHK__CHECK_INCONSIST_ACTION__CIA_IGNORE:
+		/* Report the inconsistency without repair. */
+		cbk->cb_statistics.cs_ignored++;
+		break;
+	default:
+		/*
+		 * If the specified action is not applicable to the inconsistency,
+		 * then switch to interaction mode for the decision from admin.
+		 *
+		 * Fall through.
+		 */
+	case CHK__CHECK_INCONSIST_ACTION__CIA_INTERACT:
+		options[0] = CHK__CHECK_INCONSIST_ACTION__CIA_DISCARD;
+		options[1] = CHK__CHECK_INCONSIST_ACTION__CIA_IGNORE;
+		option_nr = 2;
+		break;
+	}
+
+report:
+	cru.cru_gen = cbk->cb_gen;
+	cru.cru_cla = cla;
+	cru.cru_act = option_nr != 0 ? CHK__CHECK_INCONSIST_ACTION__CIA_INTERACT : act;
+	cru.cru_target = 0;
```

Comment:
doing this will save a lot of works:
struct chk_report_unit.   cru = {0};

Reply:
That will be simplified in subsequent patch.